Use TypeScript project references

ismail-syed opened the pull request:
## Description

Closes https://github.com/Shopify/quilt/issues/857

This PR adds Typescript [Project References](https://www.typescriptlang.org/docs/handbook/project-references.html) to quilt. In doing so we aim to see faster CI builds and faster local rebuilds. 

### Faster CI builds

* Previously [~19mins](https://travis-ci.org/Shopify/quilt/builds/611064917?utm_source=github_status&utm_medium=notification) CI runs
* Currently in this branch I'm getting [~14mins](https://travis-ci.org/Shopify/quilt/builds/611484018?utm_source=github_status&utm_medium=notification)

### Faster local rebuilds

<img width="642" alt="Screen Shot 2019-11-13 at 1 05 37 PM" src="https://user-images.githubusercontent.com/6130700/68791217-d9b84200-0616-11ea-90a7-4d1864b3c5c0.png">

ismail-syed commented on packages/address-consts/package.json:
```diff
@@ -6,10 +6,6 @@
   "sideEffects": false,
-  "scripts": {
-    "build": "tsc --p tsconfig.build.json",
-    "prepublishOnly": "yarn run build"
```
I'm not 100% sure about this, but I don't believe we need these `prepublishOnly` hooks anymore. sewing-kit monorepo packages don't include any of these so I presume it's not necessary.

TheMallen commented on packages/tsconfig.json:
```diff
@@ -0,0 +1,71 @@
+    {"path": "./semaphore"},
+    {"path": "./sewing-kit-koa"},
+    {"path": "./useful-types"},
+    {"path": "./with-env"}
```
It sucks that we have to manually maintain a list like this. It doesn't have to be part of this PR but at the least we should add a task to include adding an entry here in the new package generator, or at least have a separate script that gets run after like we do with adding to the main readme.

ismail-syed replied:
I've added a [test](https://github.com/Shopify/quilt/pull/962/files#diff-09d775e5b319188b96e0907456993c34) to cover for now.

TheMallen commented on packages/react-tracking-pixel/tsconfig.json:
```diff
@@ -0,0 +1,16 @@
+    "./src/**/*.tsx"
+  ],
+  "exclude": ["**/*.test.ts", "**/*.test.tsx"],
+  "references": [{"path": "../react-html"}]
```
We already have issues with remembering to add dependencies to `package.json` when they are local packages. We are going to want to have some way of statically enforcing that dependencies are tracked properly, whether through a linter, a custom script, a repo-level test, or whatever.

ismail-syed replied:
The compiler does an ok-ish good job at checking this. If a project is missing an explicit reference here then the build fails. That said, I've seen cases of it passing without an explicit reference. I'll open up an issue.

TheMallen commented on script.js:
```diff
@@ -0,0 +1,17 @@
+const {resolve, basename} = require('path');
```
I don't think this was meant to be left in

ismail-syed replied:
Nope, removing

TheMallen commented on packages/react-async/src/Prefetcher.tsx:
```diff
@@ -118,11 +118,13 @@ class ConnectedPrefetcher extends React.PureComponent<Props, State> {
     if (
       timeout != null &&
       urlsEqual(closestUrl, timeoutUrl) &&
+      // @ts-ignore
```
why this

ismail-syed replied:
Couple ts-ignores I forgot to address.

TheMallen commented on packages/react-i18n-universal-provider/src/I18nUniversalProvider.tsx:
```diff
@@ -26,6 +26,7 @@ export function I18nUniversalProvider({
   const [serialized, Serialize] = useSerialized<Serialized>('i18n');
 
   const i18nDetails: I18nDetails = {
+    // @ts-ignore
```
why tho

TheMallen commented on packages/react-self-serializers/src/I18n.tsx:
```diff
@@ -0,0 +1,62 @@
+import React, {useRef} from 'react';
+
+import {useSerialized, useHtmlAttributes} from '@shopify/react-html';
```
what's going on with these added typescript files? do you just need to rebase?

ismail-syed replied:
Yes, needed a rebase.

TheMallen commented on packages/graphql-testing/src/links/mocks.ts:
```diff
@@ -7,7 +7,7 @@ export class MockLink extends ApolloLink {
   }
 
-  request(operation: Operation) {
+  request(operation: Operation): Observable<any> {
```
is `any` right here?

ismail-syed replied:
I spent a bit of time trying to address these. The changes were significant in some cases. I'd prefer having those changes in a subsequent PR if that's ok with you.

TheMallen commented on packages/react-import-remote/src/hooks.ts:
```diff
@@ -103,9 +103,8 @@ export function useImportRemote<Imported = unknown>(
-          idleCallbackHandle.current = (window as WindowWithRequestIdleCallback).requestIdleCallback(
-            loadRemote,
-          );
+          idleCallbackHandle.current = (window as WindowWithRequestIdleCallback &
```
nit: maybe we should add `type ExtendedWindow<T> = window & typeof globalThis & T` to `@shopify/useful-types` and use it for all these weird window extensiony things

ismail-syed replied:
Added to useful-types

GoodForOneFare commented on yarn.lock:
```diff
@@ -11097,6 +11097,11 @@ typescript-estree@2.1.0:
     lodash.unescape "4.0.1"
     semver "5.5.0"
 
+typescript@^3.6.2:
```
Can the `@typescript@~3.2.1` (line 11105) be removed?

ismail-syed replied:
Removed

GoodForOneFare commented on package.json:
```diff
@@ -4,14 +4,14 @@
     "check": "lerna run check",
     "release": "lerna publish",
-    "clean": "rimraf './packages/*/dist/**/*.{js,d.ts}'",
+    "clean": "rimraf './packages/*/dist/**/*.{js,d.ts}' './packages/*/*.tsbuildinfo' './packages/*.tsbuildinfo'",
```
I know `dist/src` is annoying, but it's unavoidable for most projects.  So I'd prefer standardising on that, even for projects that currently use `dist`.

GoodForOneFare replied:
More importantly, how are you planning to have `tsbuildinfo` files cached in CI?  I'm still at the top of the PR, so I may be missing the magic that makes it happen.

ismail-syed replied:
> So I'd prefer standardising on that, even for projects that currently use dist

I've updated all the monorepo tsconfigs to reflect this as sewing-kit does. (`'.'` as the `baseUrl` and `rootDir`). Also changed the number of imports from `./dist/foo` to `./dist/src/foo`.

ismail-syed replied:
> how are you planning to have tsbuildinfo files cached in CI?

I tried getting Travis to only cache the `tsbuildinfo` files, but I couldn't find any docs or working examples of caching files. Only directories. How does sewing-kit cache `tsbuildinfo` files? 

In Travis, what if we move all the `tsbuildinfo`s to a `tmp_build_cache` directory of some sort after a build and dump them back into their respective `packages/*` paths before a subsequent build starts.

GoodForOneFare commented on packages/useful-types/tsconfig.json:
```diff
@@ -0,0 +1,15 @@
+{
+  "extends": "../tsconfig_base.json",
+  "compileOnSave": false,
```
`compileOnSave` should be in the base config, and deleted from all others.

GoodForOneFare commented on packages/useful-types/tsconfig.json:
```diff
@@ -0,0 +1,15 @@
+    "rootDir": "./src"
+  },
+  "include": [
+    "../../config/typescript/**/*",
```
What is this pulling in?

ismail-syed replied:
Doesn't seem like anything. I wrote a script that converted all `tsconfig.build.json
` => `tsconfig.json` and updated the extended field. Looking at the history of this file this must have been included from the generators.

GoodForOneFare commented on packages/useful-types/tsconfig.json:
```diff
@@ -0,0 +1,15 @@
+  "include": [
+    "../../config/typescript/**/*",
+    "./src/**/*.ts",
+    "./src/**/*.tsx"
```
Why does this need to process tsx files?

GoodForOneFare commented on packages/useful-types/tsconfig.json:
```diff
@@ -0,0 +1,15 @@
+    "./src/**/*.ts",
+    "./src/**/*.tsx"
+  ],
+  "exclude": ["**/*.test.ts", "**/*.test.tsx"]
```
There are no test files, so 🔥

GoodForOneFare commented on packages/with-env/tsconfig.json:
```diff
@@ -0,0 +1,15 @@
+    "rootDir": "./src"
+  },
+  "include": [
+    "../../config/typescript/**/*",
```
Same question about pulling in `config` and tsx files as above.

GoodForOneFare commented on packages/tsconfig_base.json:
```diff
@@ -22,8 +26,8 @@
   },
-  "include": ["./**/*.ts", "./**/*.tsx"],
-  "exclude": ["./node_modules"]
+  "include": ["../config/typescript/**/*.d.ts"]
```
Does it make sense to have an include here?  I thought all projects should specify their own includes?

ismail-syed replied:
This should be removed

ismail-syed commented:
Thanks for all the feedback folks. I'll get back around to addressing these comments sometime this week. Metrics tracking works is higher on my list of but I'll be been pushing this in-between cycles.

ismail-syed commented:
I've rebased the PR and addressed all the comments. Ready for another set of 👀 @GoodForOneFare @TheMallen
